src/qengine/state/opencl.cpp

```diff
@@ -10,80 +10,90 @@
 // See LICENSE.md in the project root or https://www.gnu.org/licenses/gpl-3.0.en.html
 // for details.
 
+#include <memory>
+
 #include "oclengine.hpp"
 #include "qengine_opencl.hpp"
 
 namespace Qrack {
 
 #define CMPLX_NORM_LEN 5
 
-void QEngineOCL::InitOCL()
+void QEngineOCL::SetDevice(const int& dID)
 {
-    clObj = OCLEngine::Instance();
+    if ((dID >= clObj->GetDeviceCount()) || (dID < -1)) {
+        throw "Invalid OpenCL device selection";
+    } else {
+        deviceID = -1;
+    }
+    device_context = clObj->GetDeviceContextPtr(deviceID);
```

Comment:
Is `clObj` a member variable of QEngineOCL?  It probably shouldn’t be, since it’s used very seldom it should probably be a OCLEngine::Instance() call each time it’s used.

Reply:
Removed as a member variable.